src/main/java/com/tabnine/inline/InlineCompletionHandler.java

```diff
@@ -125,14 +144,21 @@ private void rerenderCompletion(
       int offset,
       long modificationStamp,
       @NotNull CompletionAdjustment completionAdjustment) {
-    if (shouldCancelRendering(editor, modificationStamp, offset)) {
-      return;
-    }
-    if (shouldRemovePopupCompletions(completionAdjustment)) {
-      completions.removeIf(completion -> !completion.isSnippet());
-    }
-    showInlineCompletion(
-        editor, completions, offset, (completion) -> afterCompletionShown(completion, editor));
+    ApplicationManager.getApplication()
```

Comment:
doesn't this affect completions rendering? you wait for another AWT loop before rendering

Reply:
`executeUIThreadWithDelay` already does that by scheduling a non-UI thread that in his turn, will invoke a UI thread.
I changed the behaviour to schedule a non-UI thread and then wrap the rendering function with a UI thread.
this is the flow:
1. document change event occurred
2. invoking a UI thread which will be invoked after all document changed events are fired (can be more than one for a single change)
3. invoking a non-UI thread for fetching completions
4. if debounce time is 0, execute the rendering function, which is wrapped by a UI thread. if denounce time is greater than 0 - schedule a non-UI thread that will invoke the rendering function.

I think it's safer to wrap the rendering method with a UI thread from the inside rather than manage it from the outside, which can lead to rendering in the non-UI thread

Reply:
I'm worried that since you're waiting a total of 2 UI event loops, you'll be messing with the user's actions in a buggy manner. For example, what happens if the user types / switches tab / etc.  between these 2 loop iterations you're waiting? it can lead to some messy edge cases IMO

Reply:
you're right, and this is exactly the reason why we add this:

https://github.com/codota/tabnine-intellij/pull/450/files#diff-d6849f3d63babae30d241aa3112fdab621a3a77b2256ac66b4d51f6e87f02c8fR164

this is a validation of the offset and the modificationStamp between the first loop and the second loop so we ensure there is no change between them

Reply:
anyway, you must invoke a new thread to fetch the completions, and then invoke a new UI thread for rendering, so there is no way to sneak out of that